bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go

```diff
@@ -107,6 +108,17 @@ type KubeadmConfigSpec struct {
 	Ignition *IgnitionSpec `json:"ignition,omitempty"`
 }
 
+// Validate ensures the KubeadmConfigSpec is valid.
+func (c *KubeadmConfigSpec) Validate(pathPrefix *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+
+	allErrs = append(allErrs, validateFiles(c, pathPrefix)...)
+	allErrs = append(allErrs, validateUsers(c, pathPrefix)...)
+	allErrs = append(allErrs, validateIgnition(c, pathPrefix)...)
+
+	return allErrs
+}
```

Comment:
I wouldn't have expected a validate function to be a member of `KubeadmConfigSpec`, I don't recall doing this for others?

Reply:
I have to take a closer look but KubeadmConfigSpec is a very special case as we need the validation in KCP as well (IIRC). We also have to use the default func in KCP.

Reply:
As @sbueringer is mentioning, we are using the Validate function all over the place.
I do think that in this case, the Validate function being a member of `KubeadmConfigSpec` is not a bad idea. On the other hand, we would also need to move `validate{Files, Users, Ignition}` to some other accessible file (not internal).

What I have right now locally is that I just copied this code from `bootstrap/kubeadm` to also `controlplane/kubeadm` which uses it. I did the same thing for the defaulting function in KCP. 

I will try to think about some more sensible way to do this. If you have any ideas, please let me know.

Reply:
I would be in favor of keeping the `DefaultKubeadmConfigSpec` and `(c *KubeadmConfigSpec) validate(name string)` funcs in the v1beta1 API package (and just move them from the _webhook to the _types file) (also includes the funcs that validate is calling)

I think this should allow us to move the webhooks without having to do complex refactorings

Reply:
The goal of this issue/PR is mainly to get rid of the controller-runtime dependencies, we can do that without touching these functions.

Reply:
Sure, if that drops the CR dep then lets do that for now and see how we get on

Reply:
Wouldn't it be better to move the functions that Validate is calling to move somewhere else than the _types file, just to not pollute the file that much?

Reply:
Fine for me